src/com/facebook/buck/cxx/HeaderSymlinkTreeWithHeaderMap.java

```diff
@@ -72,10 +120,44 @@ public SourcePath getSourcePathToOutput() {
       // aligning in order to get this to work. May we find peace in another life.
       headerMapEntries.put(key, buckOut.relativize(getRoot().resolve(key)));
     }
-    return ImmutableList.<Step>builder()
+    ImmutableList.Builder<Step> builder = ImmutableList.<Step>builder()
         .addAll(super.getBuildSteps(context, buildableContext))
-        .add(new HeaderMapStep(getProjectFilesystem(), headerMapPath, headerMapEntries.build()))
-        .build();
+        .add(new HeaderMapStep(getProjectFilesystem(), headerMapPath, headerMapEntries.build()));
+
+    if (shouldCreateModule) {
+      Optional<String> umbrellaHeader = getUmbrellaHeader(getBuildTarget().getShortName());
```

Comment:
Implicitly use `targetName.h` as the umbrella header file of a module if exists (which is the way xcode structures a framework), otherwise we use umbrella directory (http://clang.llvm.org/docs/Modules.html#umbrella-directory-declaration). 

Another way for doing this more explicitly is that we can add an extra argument for `cxx_*`, `apple_*` rules to point to the umbrella header.